### DISCUSSION TITLE
how to make use of extra memory resources?

### DISCUSSION THREAD

**Discussion by akcopy** [milvus-io/milvus#32233] [Q&A and General discussion]
I'm currently running milvus standalone on a single server. After lots of tweaking I found that I have a lot of additional memory (more than what I need for the tasks). Is it possible to have multiple in memory replicas to boost the QPS of my system to speed it up? Or are there any other tricks which I can use the additional memory? The only way I can think of is to have 2 completely separate milvus instances running but that seems to be a pain to setup and this might not even be optimal since there is only 1 CPU and 1 GPU (they would be competing against each other for resources).

**Comment by xiaofan-luan** [comment]
> I'm currently running milvus standalone on a single server. After lots of tweaking I found that I have a lot of additional memory (more than what I need for the tasks). Is it possible to have multiple in memory replicas to boost the QPS of my system to speed it up? Or are there any other tricks which I can use the additional memory? The only way I can think of is to have 2 completely separate milvus instances running but that seems to be a pain to setup and this might not even be optimal since there is only 1 CPU and 1 GPU (they would be competing against each other for resources).

Doesn't really help your performance if cpu core doesn't increase.

**Reply by akcopy** [reply]
that is what I suspect too. Is there other ways to maximize performance just with milvus standalone? Or is my only option to run milvus cluster and add multiple query nodes on just 1 server?

**Comment by xiaofan-luan** [comment]
how many data, how many QPS you are targeting at? and how many cores are there in your cluster?

1. using HNSW can improve the performance
2. if you already use hnsw, increase the segment size can improve the performance

**Reply by akcopy** [reply]
my databases are quite small - 7 million vectors of 1024 dim on a GPU_IVF_PQ index and another database of around 300,000 vectors of 1024 dim on a SCANN index. I run this on a 32-core machine with a RTX 3090 (24GB) and 256GB ram, and I realized that it appears to be a bit overkill for the application. But if I can boost the performance significantly this will save me buying additional servers in the future?

**Comment by yhmo** [comment]
I think you can try this approach:
1. create two collections, they have the same data (7 million vectors of 1024 dim).
2. collection_A uses GPU_IVF_PQ, the index is loaded into GPU memory, search requests to this collection are executed by GPU.
3. collection_B uses HNSW index, the index is loaded into CPU memory, search requests to this collection are executed by CPU.
4. you can make a request balancer in client-side, half of the requests are sent to collection_A, another half are sent to collection_B 

The target is to exhaust the hardware resources to get higher QPS. More things you can try:

If you deployed a Milvus cluster on the machine, you can use "replicas" to load the collection_B to multiple copies in memory to get higher QPS: 
https://milvus.io/docs/v2.3.x/load_collection.md. (Note that the number of query nodes must be equal or larger than replica_number).

Since the RAM is 256GB, the collection's raw data size is 7M * 1024 * 4bytes = 28GB, you can set "replicas" to 4 ~6(requires 4 ~ 6 query nodes), it will occupy 112GB ~ 168GB RAM. Milvus automatically balances the search requests to different replicas.
`collection_B.load(replica_number=6)`
